File: backend/app/api/agents.py
import traceback
from flask import request, jsonify, current_app
from . import agents_bp
from ..models.agent import Agent
from ..services.agent_store import AgentStore
from ..services.llm import call_llm

def _store(): return AgentStore(current_app.config['AF_CFG'].AGENTS_FILE)

@agents_bp.get('/')
def list_agents():
    return jsonify({'success':True,'data':[a.to_dict() for a in _store().list()]})

@agents_bp.post('/')
def create():
    try:
        d = request.get_json() or {}
        if not d.get('name') or not d.get('provider'):
            return jsonify({'success':False,'error':'name and provider required'}), 400
        a = Agent(name=d['name'],provider=d['provider'],api_key=d.get('api_key',''),
                  model=d.get('model',''),base_url=d.get('base_url',''),
                  role=d.get('role','specialist'),description=d.get('description',''),
                  enabled=d.get('enabled',True))
        _store().save(a); return jsonify({'success':True,'data':a.to_dict()}), 201
    except Exception as e:
        return jsonify({'success':False,'error':str(e),'tb':traceback.format_exc()}), 500

@agents_bp.get('/<aid>')
def get(aid):
    a = _store().get(aid)
    if not a: return jsonify({'success':False,'error':f'Not found: {aid}'}), 404
    return jsonify({'success':True,'data':a.to_dict()})

@agents_bp.put('/<aid>')
def update(aid):
    try:
        s = _store(); a = s.get(aid)
        if not a: return jsonify({'success':False,'error':f'Not found: {aid}'}), 404
        d = request.get_json() or {}
        for f in ('name','provider','model','base_url','role','description','enabled'):
            if f in d: setattr(a,f,d[f])
        if d.get('api_key'): a.api_key = d['api_key']
        s.save(a); return jsonify({'success':True,'data':a.to_dict()})
    except Exception as e:
        return jsonify({'success':False,'error':str(e)}), 500

@agents_bp.delete('/<aid>')
def delete(aid):
    ok = _store().delete(aid)
    if not ok: return jsonify({'success':False,'error':f'Not found: {aid}'}), 404
    return jsonify({'success':True})

@agents_bp.post('/<aid>/test')
def test(aid):
    a = _store().get(aid)
    if not a: return jsonify({'success':False,'error':f'Not found: {aid}'}), 404
    try:
        r = call_llm(a,'You are a helpful assistant.','Reply with exactly: AssetFlow OK')
        return jsonify({'success':True,'response':r[:200]})
    except Exception as e:
        return jsonify({'success':False,'error':str(e)}), 400

File: backend/routes/agents.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const { getAgents, upsertAgent, deleteAgent, getAgent } = require('../services/storage');
const { callLLM } = require('../services/llm');

const router = express.Router();

// GET /api/agents
router.get('/', async (req, res) => {
  try {
    const agents = await getAgents();
    // Never return api keys to frontend
    res.json(agents.map(sanitize));
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// POST /api/agents
router.post('/', async (req, res) => {
  try {
    const { name, provider, apiKey, model, baseUrl, role, description } = req.body;
    if (!name || !provider) return res.status(400).json({ error: 'name and provider required' });

    const agent = {
      id: uuidv4(),
      name,
      provider,
      apiKey: apiKey || '',
      model: model || '',
      baseUrl: baseUrl || '',
      role: role || 'specialist',
      description: description || '',
      enabled: true,
      createdAt: new Date().toISOString(),
    };

    await upsertAgent(agent);
    res.json(sanitize(agent));
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// PUT /api/agents/:id
router.put('/:id', async (req, res) => {
  try {
    const existing = await getAgent(req.params.id);
    if (!existing) return res.status(404).json({ error: 'Agent not found' });

    const updated = { ...existing, ...req.body, id: existing.id, createdAt: existing.createdAt };
    await upsertAgent(updated);
    res.json(sanitize(updated));
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// DELETE /api/agents/:id
router.delete('/:id', async (req, res) => {
  try {
    const deleted = await deleteAgent(req.params.id);
    if (!deleted) return res.status(404).json({ error: 'Agent not found' });
    res.json({ success: true });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// POST /api/agents/:id/test - Test agent connectivity
router.post('/:id/test', async (req, res) => {
  try {
    const agent = await getAgent(req.params.id);
    if (!agent) return res.status(404).json({ error: 'Agent not found' });

    const response = await callLLM(
      agent,
      'You are a helpful assistant.',
      'Say "AssetFlow connection successful" and nothing else.'
    );

    res.json({ success: true, response: response?.slice(0, 200) });
  } catch (err) {
    res.status(400).json({ success: false, error: err.message });
  }
});

function sanitize(agent) {
  const { apiKey, ...safe } = agent;
  return { ...safe, hasApiKey: !!(apiKey && apiKey.length > 0) };
}

module.exports = router;
